File: C++/knight-probability-in-chessboard.cpp
// Time:  O(k * n^2)
// Space: O(n^2)

class Solution {
public:
    double knightProbability(int N, int K, int r, int c) {
        static const vector<pair<int, int>> directions = 
            {{ 1, 2}, { 1, -2}, { 2, 1}, { 2, -1},
             {-1, 2}, {-1, -2}, {-2, 1}, {-2, -1}};
        vector<vector<vector<double>>> dp(2, vector<vector<double>>(N, vector<double>(N, 1.0l)));
        for (int step = 1; step <= K; ++step) {
            for (int i = 0; i < N; ++i) {
                for (int j = 0; j < N; ++j) {
                    dp[step % 2][i][j] = 0;
                    for (const auto& direction : directions) {
                        auto rr = i + direction.first;
                        auto cc = j + direction.second;
                        if (0 <= cc && cc < N && 0 <= rr && rr < N) {
                            dp[step % 2][i][j] += 0.125l * dp[(step - 1) % 2][rr][cc];
                        }
                    }
                }
            }
        }
        return dp[K % 2][r][c];
    }
};

File: Python/knight-probability-in-chessboard.py
# Time:  O(k * n^2)
# Space: O(n^2)
   
class Solution(object):
    def knightProbability(self, N, K, r, c):
        """
        :type N: int
        :type K: int
        :type r: int
        :type c: int
        :rtype: float
        """
        directions = \
            [[ 1, 2], [ 1, -2], [ 2, 1], [ 2, -1], \
             [-1, 2], [-1, -2], [-2, 1], [-2, -1]];
        dp = [[[1 for _ in xrange(N)] for _ in xrange(N)] for _ in xrange(2)]
        for step in xrange(1, K+1):
            for i in xrange(N):
                for j in xrange(N):
                    dp[step%2][i][j] = 0
                    for direction in directions:
                        rr, cc = i+direction[0], j+direction[1]
                        if 0 <= cc < N and 0 <= rr < N:
                            dp[step%2][i][j] += 0.125 * dp[(step-1)%2][rr][cc];

        return dp[K%2][r][c]
